src/stan/services/optimize/laplace_sample.hpp

```diff
@@ -70,54 +70,53 @@ void laplace_sample(const Model& model, const Eigen::VectorXd& theta_hat,
   math::internal::finite_diff_hessian_auto(log_density_fun, theta_hat, log_p,
                                            grad, hessian);
   if (refresh > 0) {
-    logger.info(log_density_msgs);
-    logger.info("\n");
+    if (log_density_msgs.peek() != std::char_traits<char>::eof())
+      logger.info(log_density_msgs);
   }
 
   // calculate Cholesky factor and inverse
   interrupt();
   if (refresh > 0) {
     logger.info("Calculating inverse of Cholesky factor");
-    logger.info("\n");
   }
   Eigen::MatrixXd L_neg_hessian = (-hessian).llt().matrixL();
   interrupt();
   Eigen::MatrixXd inv_sqrt_neg_hessian = L_neg_hessian.inverse().transpose();
   interrupt();
   Eigen::MatrixXd half_hessian = 0.5 * hessian;
 
-  // generate draws and output to sample writer
   if (refresh > 0) {
     logger.info("Generating draws");
-    logger.info("\n");
   }
+  // generate draws
+  std::stringstream refresh_msg;
   boost::ecuyer1988 rng = util::create_rng(random_seed, 0);
   Eigen::VectorXd draw_vec;  // declare draw_vec, msgs here to avoid re-alloc
   for (int m = 0; m < draws; ++m) {
     interrupt();  // allow interpution each iteration
     if (refresh > 0 && m % refresh == 0) {
-      logger.info("iteration: ");
-      logger.info(std::to_string(m));
-      logger.info("\n");
+      refresh_msg << "iteration: " << std::to_string(m);
+      logger.info(refresh_msg);
+      refresh_msg.str(std::string());
     }
     Eigen::VectorXd z(num_unc_params);
     for (int n = 0; n < num_unc_params; ++n) {
       z(n) = math::std_normal_rng(rng);
     }
-
     Eigen::VectorXd unc_draw = theta_hat + inv_sqrt_neg_hessian * z;
-    double log_p = log_density_fun(unc_draw).val();
-    Eigen::VectorXd diff = unc_draw - theta_hat;
-    double log_q = diff.transpose() * half_hessian * diff;
-
-    std::stringstream msgs;
-    model.write_array(rng, unc_draw, draw_vec, include_tp, include_gq, &msgs);
+    std::stringstream write_array_msgs;
+    model.write_array(rng, unc_draw, draw_vec, include_tp, include_gq,
+                      &write_array_msgs);
     if (refresh > 0) {
```

Comment:
This should be `refresh > 0 && write_array_msgs.peek() != std::char_traits<char>::eof()` in a single condition---no need to nest.   (Unless there's other stuff in the output---it's hard to see in the diff.)

Reply:
agree